web/packages/teleport/src/Account/ManageDevices/AddAuthDeviceWizard/AddAuthDeviceWizard.tsx

```diff
@@ -246,67 +284,198 @@ function CreatePasskeyStep({
   );
 }
 
+function CreateMfaBox({
+  auth2faType,
+  newMfaDeviceType,
+  privilegeToken,
+  onNewMfaDeviceTypeChange,
+}: {
+  auth2faType: Auth2faType;
+  newMfaDeviceType: Auth2faType;
+  privilegeToken: string;
+  onNewMfaDeviceTypeChange(o: Auth2faType): void;
+}) {
+  const mfaOptions = createMfaOptions({
+    auth2faType,
+    required: true,
+  }).map((o: MfaOption) =>
+    // Be more specific about the WebAuthn device type (it's not a passkey).
+    o.value === 'webauthn' ? { ...o, label: 'Hardware Device' } : o
+  );
+
+  return (
+    <>
+      Multi-factor type
+      <RadioGroup
+        name="mfaOption"
+        options={mfaOptions}
+        value={newMfaDeviceType}
+        flexDirection="row"
+        gap={3}
+        onChange={o => {
+          onNewMfaDeviceTypeChange(o as Auth2faType);
+        }}
+      />
+      {newMfaDeviceType === 'otp' && (
+        <QrCodeBox privilegeToken={privilegeToken} />
+      )}
+    </>
+  );
+}
+
+function QrCodeBox({ privilegeToken }: { privilegeToken: string }) {
+  const fetchQrCodeAttempt = useAttempt('');
+  const [qrCode, setQrCode] = useState('');
+
+  useEffect(() => {
+    fetchQrCodeAttempt.run(() =>
+      auth
+        .createMfaRegistrationChallenge(privilegeToken, 'totp')
+        .then(res => setQrCode(res.qrCode))
+    );
+  }, []);
```

Comment:
Just a nit, `useAsync` could replace these two calls:

<details>
<summary>Diff</summary>

Apply by copying and then `pbpaste | git apply`

```diff
diff --git a/web/packages/teleport/src/Account/ManageDevices/AddAuthDeviceWizard/AddAuthDeviceWizard.tsx b/web/packages/teleport/src/Account/ManageDevices/AddAuthDeviceWizard/AddAuthDeviceWizard.tsx
index 4afd5dbdd6..158fb8e9d2 100644
--- a/web/packages/teleport/src/Account/ManageDevices/AddAuthDeviceWizard/AddAuthDeviceWizard.tsx
+++ b/web/packages/teleport/src/Account/ManageDevices/AddAuthDeviceWizard/AddAuthDeviceWizard.tsx
@@ -31,6 +31,7 @@ import React, { useState, useEffect, FormEvent } from 'react';
 import FieldInput from 'shared/components/FieldInput';
 import Validation, { Validator } from 'shared/components/Validation';
 import { requiredField } from 'shared/components/Validation/rules';
+import { useAsync } from 'shared/hooks/useAsync';
 import useAttempt from 'shared/hooks/useAttemptNext';
 import { Auth2faType } from 'shared/services';
 import createMfaOptions, { MfaOption } from 'shared/utils/createMfaOptions';
@@ -324,15 +325,12 @@ function CreateMfaBox({
 }
 
 function QrCodeBox({ privilegeToken }: { privilegeToken: string }) {
-  const fetchQrCodeAttempt = useAttempt('');
-  const [qrCode, setQrCode] = useState('');
+  const [fetchQrCodeAttempt, fetchQrCode] = useAsync((privilegeToken: string) =>
+    auth.createMfaRegistrationChallenge(privilegeToken, 'totp')
+  );
 
   useEffect(() => {
-    fetchQrCodeAttempt.run(() =>
-      auth
-        .createMfaRegistrationChallenge(privilegeToken, 'totp')
-        .then(res => setQrCode(res.qrCode))
-    );
+    fetchQrCode(privilegeToken);
   }, []);
 
   return (
@@ -346,10 +344,10 @@ function QrCodeBox({ privilegeToken }: { privilegeToken: string }) {
       `}
     >
       <Flex height="168px" justifyContent="center" alignItems="center">
-        {fetchQrCodeAttempt.attempt.status === 'processing' && <Indicator />}
-        {fetchQrCodeAttempt.attempt.status === 'success' && (
+        {fetchQrCodeAttempt.status === 'processing' && <Indicator />}
+        {fetchQrCodeAttempt.status === 'success' && (
           <Image
-            src={`data:image/png;base64,${qrCode}`}
+            src={`data:image/png;base64,${fetchQrCodeAttempt.data.qrCode}`}
             height="100%"
             style={{
               boxSizing: 'border-box',

```
</details>

Reply:
@ravicious That's a nice trick, will use it. Though I think that the `privilegeToken` parameter should come from the function closure, not a parameter.

Reply:
It totally could, but this way the function passed to `useAsync` doesn't take any extra dependencies. This doesn't matter here, but I got used to writing functions like that whenever I had to construct them with `useAsync(useCallback(…))`.

Fingers crossed that the React Compiler gets rid of nuances like that completely.